src/librustc_typeck/check/pat.rs

```diff
@@ -668,14 +681,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
 
         // Type-check the path.
-        let pat_ty = self.instantiate_value_path(segments, opt_ty, res, pat.span, pat.hir_id).0;
-        self.demand_suptype(pat.span, expected, pat_ty);
+        let (pat_ty, pat_res) =
+            self.instantiate_value_path(segments, opt_ty, res, pat.span, pat.hir_id);
+
+        self.demand_suptype_diag(pat.span, expected, pat_ty).map(|e| {
+            self.emit_bad_pat_path(e, pat.span, res, pat_res, qpath, segments, parent_pat)
+        });
         pat_ty
     }
 
+    fn emit_bad_pat_path(
+        &self,
+        mut e: DiagnosticBuilder<'_>,
+        pat_span: Span,
+        res: Res,
+        pat_res: Res,
+        qpath: &hir::QPath<'_>,
+        segments: &'b [hir::PathSegment<'b>],
+        parent_pat: Option<&Pat<'_>>,
+    ) {
+        let path = hir::print::to_string(self.tcx.hir(), |s| s.print_qpath(qpath, false));
+        if let Some(span) = self.tcx.hir().res_span(pat_res) {
+            e.span_label(span, &format!("{} defined here", res.descr()));
+            if let [hir::PathSegment { ident, .. }] = &*segments {
+                e.span_label(
+                    pat_span,
+                    &format!(
+                        "`{}` is interpreted as {} {}, not a new binding",
+                        path,
+                        res.article(),
+                        res.descr(),
+                    ),
+                );
+                if let Some(Pat { kind: hir::PatKind::Struct(..), .. }) = parent_pat {
```

Comment:
```suggestion
                let (msg, sugg) = match parent_pat {
```
extract out the common parts, including `ident.as_str().to_lowercase()`.